src/Condition.js

```diff
@@ -1,59 +1,75 @@
 import has from 'lodash/has';
 import isSchema from './util/isSchema';
 
-function callOrConcat(schema) {
-  if (typeof schema === 'function') return schema;
+function wrapCusomFn(fn) {
+  return function(...args) {
+    args.pop();
+    return fn.apply(this, args);
+  };
+}
+
+function makeConditionPredicate(is) {
+  if (typeof is === 'function') return is;
 
-  return base => base.concat(schema);
+  return (...values) => values.every(value => value === is);
 }
 
-class Conditional {
-  constructor(refs, options) {
-    let { is, then, otherwise } = options;
+function makeBranchCustomizer(branch) {
+  if (!branch) return () => undefined;
 
-    this.refs = [].concat(refs);
+  if (typeof branch === 'function') return schema => branch(schema);
 
-    then = callOrConcat(then);
-    otherwise = callOrConcat(otherwise);
+  return (schema, options) => schema.concat(branch.resolve(options));
+}
 
-    if (typeof options === 'function') this.fn = options;
-    else {
-      if (!has(options, 'is'))
-        throw new TypeError('`is:` is required for `when()` conditions');
+function makeFn(options) {
+  if (typeof options === 'function') return wrapCusomFn(options);
 
-      if (!options.then && !options.otherwise)
-        throw new TypeError(
-          'either `then:` or `otherwise:` is required for `when()` conditions',
-        );
+  if (!has(options, 'is'))
+    throw new TypeError('`is:` is required for `when()` conditions');
 
-      let isFn =
-        typeof is === 'function'
-          ? is
-          : (...values) => values.every(value => value === is);
+  if (!options.then && !options.otherwise)
+    throw new TypeError(
+      'either `then:` or `otherwise:` is required for `when()` conditions',
+    );
 
-      this.fn = function(...values) {
-        let currentSchema = values.pop();
-        let option = isFn(...values) ? then : otherwise;
+  let { is, then, otherwise } = options;
 
-        return option(currentSchema);
-      };
-    }
-  }
+  is = makeConditionPredicate(is);
+  then = makeBranchCustomizer(then);
+  otherwise = makeBranchCustomizer(otherwise);
```

Comment:
again lets remove `makeBranchCustomizer` (as well as `makeConditionPredicate`) and move it inline to line 47

Reply:
Any reason? That way it is more performant.

Reply:
its mild indirection for an unverified perf benefit. I'm skeptical even that it could/would be, the above creates a bunch of unneeded closures and functions.

Reply:
Done.